torchrl/envs/libs/meltingpot.py

```diff
@@ -0,0 +1,594 @@
+                    },
+                )
+            )
+            observation_specs.append(
+                CompositeSpec(
+                    {
+                        "observation": torchrl_agent_obs_specs[
+                            agent_index
+                        ]  # shape = (n_obs_per_agent,)
+                    },
+                )
+            )
+            reward_specs.append(
+                CompositeSpec({"reward": torchrl_rew_spec[agent_index]})  # shape = (1,)
+            )
+
+        # Create multi-agent specs
+        group_action_spec = torch.stack(
+            action_specs, dim=0
+        )  # shape = (n_agents_in_group, n_actions_per_agent)
+        group_observation_spec = torch.stack(
+            observation_specs, dim=0
+        )  # shape = (n_agents_in_group, n_obs_per_agent)
+        group_reward_spec = torch.stack(
+            reward_specs, dim=0
+        )  # shape = (n_agents_in_group, 1)
+        return (
+            group_observation_spec,
+            group_action_spec,
+            group_reward_spec,
+        )
+
+    def _check_kwargs(self, kwargs: Dict):
+        meltingpot = self.lib
+
+        if "env" not in kwargs:
+            raise TypeError("Could not find environment key 'env' in kwargs.")
+        env = kwargs["env"]
+        if not isinstance(env, meltingpot.utils.substrates.substrate.Substrate):
+            raise TypeError(
+                "env is not of type 'meltingpot.utils.substrates.substrate.Substrate'."
+            )
+
+    def _init_env(self):
+        # Caching
+        self.cached_full_done_spec_zero = self.full_done_spec.zero()
+
+    def _set_seed(self, seed: Optional[int]):
+        raise NotImplementedError(
+            "It is unclear how to set a seed in Meltingpot"
+            " (https://github.com/google-deepmind/meltingpot/issues/129)"
+        )
+
+    def _reset(
+        self, tensordict: Optional[TensorDictBase] = None, **kwargs
+    ) -> TensorDictBase:
+        self.num_cycles = 0
+        timestep = self._env.reset()
+        obs = timestep.observation
+
+        td = self.cached_full_done_spec_zero.clone()
+
+        for group, agent_names in self.group_map.items():
+            agent_tds = []
+            for index_in_group, agent_name in enumerate(agent_names):
+                global_index = self.agent_names_to_indices_map[agent_name]
+                agent_obs = self.observation_spec[group, "observation"][
+                    index_in_group
+                ].encode(_filter_global_state_from_dict(obs[global_index], world=False))
+                agent_td = TensorDict(
+                    source={
+                        "observation": agent_obs,
+                    },
+                    batch_size=self.batch_size,
+                    device=self.device,
+                )
+
+                agent_tds.append(agent_td)
+            agent_tds = torch.stack(agent_tds, dim=0)
+            td.update({group: agent_tds})
```

Comment:
why not `td.set(group, agent_tds)` which will be faster (no intermediate conversion to a tensordict)

Reply:
Thanks for catching this!